src/mozanalysis/metrics/__init__.py

```diff
@@ -170,3 +170,12 @@ def agg_sum(select_expr):
 def agg_any(select_expr):
     """Return the logical OR, with FALSE-filled nulls."""
     return "COALESCE(LOGICAL_OR({}), FALSE)".format(select_expr)
+
+
+def agg_histogram_mean(select_expr):
+    """Produces an expression for the mean of an unparsed histogram."""
+    return dedent(f"""\
+        SAFE_DIVIDE(
+            SUM(CAST(JSON_EXTRACT_SCALAR({select_expr}, "$.sum") AS int64)),
+            SUM((SELECT SUM(value) FROM UNNEST(`moz-fx-data-shared-prod`.udf.json_extract_histogram({select_expr}).values)))
+        )""")
```

Comment:
This will sometimes return `NULL`, won't it? Is there a way to analyse the quantity that you want to analyse that doesn't involve dropping nulls before applying any stats? E.g. counting the fraction of users above some threshold?

It's hard to rule out the possibility that dropping nulls is biasing the results in favour of one branch - and even making that check in the first place is an extra step that's easily forgotten.

So, if it's possible to achieve your analysis goals in a way that safely handles nulls without adding any footguns, then that would be preferable. If not, then we'll have to continue to live dangerously.

Reply:
> This will sometimes return NULL, won't it?

Yes.

> Is there a way to analyse the quantity that you want to analyse that doesn't involve dropping nulls before applying any stats? 

No; I think the line I want to take is that users without any e.g. DNS activity simply don't inform an analysis of DNS performance. These metrics should be applied to an analysis window spanning the entire experiment, so that users that flaked because of a poor experience are still included in the result. The user average value is a natural and useful quantity. I appreciate they're not obviously appropriate for timeseries or engagement metrics but we probably don't encode the latter as histograms very often.

Reply:
> These metrics should be applied to an analysis window spanning the entire experiment, so that users that flaked because of a poor experience are still included in the result.

Is there a way we can design things such that the set of "what is straightforward to query using mozanalysis" is as close as possible to the set of "what is a valid analysis"? Specifically, if this metric only makes sense for an analysis window starting at 0, should we try to enforce that (either weakly, with a warning, or strictly, with an error)?

And/or should we force-add a sanity metric checking the NULL rates in each branch?

Reply:
> Is there a way we can design things such that the set of "what is straightforward to query using mozanalysis" is as close as possible to the set of "what is a valid analysis"? Specifically, if this metric only makes sense for an analysis window starting at 0, should we try to enforce that (either weakly, with a warning, or strictly, with an error)?

That introduces a linkage between a metric and a concrete analysis window, which seems like it contradicts the design philosophy. I'm optimistic that review will catch egregious errors.

> And/or should we force-add a sanity metric checking the NULL rates in each branch?

If a metric is a Metric, I don't think there's anything we could add here. Do you mean the subsequent summary ops could report the frequency of NULLs?

Reply:
> That introduces a linkage between a metric and a concrete analysis window, which seems like it contradicts the design philosophy. I'm optimistic that review will catch egregious errors.

It does contradict the design philosophy, but if we have a runtime check that enforces it weakly, then I consider this _bending_ rather than breaking the philosophy. I'm less optimistic about review than you are, but you've considered my argument so I'll disagree and commit.

> If a metric is a Metric, I don't think there's anything we could add here. Do you mean the subsequent summary ops could report the frequency of NULLs?

It's a common (and in my POV dangerous) request to have metrics that return NULLs, so that the NULLs can be dropped before running stats. If we're going to accept that this evil is necessary (in some cases it probably is), then we could stick some logic somewhere in mozanalysis that checks which elements of `metrics_list` can return NULLs, and adds a binary column for each of those metrics, containing whether the column value is NULL. Yes, this is redundant information, but it's putting an important easily-forgotten quantity right in front of the user.

Reply:
I think we agree that NULLs should be acknowledged; I just think the next aggregation level is the right place to add that column.

Reply:
I can imagine mozilla DSes picking and choosing which parts of the stack they want to use - do they want to use the whole dashboarding solution? Do they want to use mozanalysis manually in a notebook to get the data, aggregate and run the stats? Do they just want to use the stats functions? Do they just want to get the data then use their own stats functions?

Where should we add the NULL-checking to be safe for the most people? Above, I thought it should go into the "get the data" phase to catch the most people, but maybe it should go into the stats functions - the bootstrap functions could automatically drop nulls then also check for bias and return the extra results??

Reply:
...or we add new bootstrap functions that drop the nulls and call the existing bootstrap functions as well as the sanity checks - which will have better properties w.r.t. consistent return types.

Reply:
Sure - something like "Add bootstrap functions that support data with NULLs"?